pkg/gadgets/trace/tcpretrans/tracer/bpf/tcpretrans.bpf.c

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: GPL-2.0
+//
+// Based on tcpretrans(8) from BCC
+//
+// Copyright 2016 Netflix, Inc.
+//
+// 14-Feb-2016   Brendan Gregg   Created this.
+// 03-Nov-2017   Matthias Tafelmeier Extended this.
+// Copyright 2023 Microsoft Corporation
+
+#include <vmlinux/vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+
+#define GADGET_TYPE_TRACING
+#include <sockets-map.h>
+
+#include "tcpretrans.h"
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+// we need this to make sure the compiler doesn't remove our struct
+const struct event *unusedevent __attribute__((unused));
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+// This struct is defined according to the following format file:
+// /sys/kernel/tracing/events/tcp/tcp_retransmit_skb/format
+struct tcp_retransmit_skb_ctx {
+	/* The first 8 bytes are not allowed to read */
+	unsigned long pad;
+
+	void *skbaddr;
+	void *skaddr;
+	int state;
+	__u16 sport;
+	__u16 dport;
+	__u8 saddr[4];
+	__u8 daddr[4];
+	__u8 saddr_v6[16];
+	__u8 daddr_v6[16];
+};
+
+// This struct is the same as struct tcphdr in vmlinux.h but with flags defined as single field instead of bitfield
+struct tcphdr_with_flags {
+	__be16 source;
+	__be16 dest;
+	__be32 seq;
+	__be32 ack_seq;
+	__u16 res1: 4;
+	__u16 doff: 4;
+	__u8 flags;
+	__be16 window;
+	__sum16 check;
+	__be16 urg_ptr;
+};
+
+static __always_inline int __trace_tcp_retrans(void *ctx, struct sock *sk, struct sk_buff *skb)
+{
+	if (sk == NULL)
+		return 0;
+
+	if (BPF_CORE_READ_BITFIELD_PROBED(sk, sk_protocol) != IPPROTO_TCP)
+		return 0;
+
+	struct inet_sock *sockp = (struct inet_sock *)sk;
```

Comment:
In most of our gadgets we try to keep variable declarations at the top. (As done in the Linux kernel), perhaps we should try to follow that style.